fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,118 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    this.getAllTodos();
+  }
+
+  getAllTodos() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+
+    fetch('/api/todos')
+      .then(data => data.json())
+      .then(todos => {
+        this.setState({ todos: todos });
+      });
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    */
+
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(todo => {
+        this.setState(state => {
+          state.todos.push(todo);
+          return { todos: state.todos };
+        });
+      });
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    let updateIndex;
+    const todoStatus = this.state.todos.filter((todo, index) => {
+      if (todo.todoId === todoId) {
+        updateIndex = index;
+      }
+      return todo.todoId === todoId;
+    });
+    let updatedTodoStatus;
+    if (todoStatus[0].isCompleted === false) {
+      updatedTodoStatus = true;
+    } else {
+      updatedTodoStatus = false;
+    }
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({ isCompleted: updatedTodoStatus })
+    })
+      .then(res => {
+        this.setState(state => {
+          state.todos[updateIndex].isCompleted = updatedTodoStatus;
```

Comment:
We never modify objects and arrays in state. You need to create a copy of the array and replace the old object with the new object.